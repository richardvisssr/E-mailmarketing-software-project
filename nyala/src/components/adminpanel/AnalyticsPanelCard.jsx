import React, { useState, useEffect } from "react";
import AlertComponent from "../alert/AlertComponent";

const AnalyticsPanelCard = (props) => {
  const socket = new WebSocket("ws://localhost:7002/socket");
  const [openedOnlineEmails, setOpenedOnlineEmails] = useState(0);
  const [unsubscribe, setUnsubscribe] = useState(0);
  const [notification, setNotification] = useState({ type: "", message: "" });
  const id = props.id;

  useEffect(() => {
    fetch(`http://localhost:3001/stats/${id}`, {})
      .then((response) => response.json())
      .then((data) => {
        const key = Object.keys(data)[0];
        const value = data[key];
        setOpenedOnlineEmails(value.opened);
        setUnsubscribe(value.unsubscribed);
      })
      .catch((error) => {
        setNotification({
          type: "error",
          message: "Er ging iets mis met het ophalen van de data",
        });
      });
  }, [id]);

  socket.addEventListener("open", (event) => {});

  socket.addEventListener("message", (event) => {
    try {
      const message = JSON.parse(event.data);
      const emailId = message.emailId;
      const currentEmailId = id; // assuming template.id is the emailId

      if (
        (message.type === "trackOnlineView" ||
          message.type === "trackUnsubscribe") &&
        emailId === currentEmailId
      ) {
        if (message.type === "trackOnlineView") {
          setOpenedOnlineEmails(message.opened);
        } else if (message.type === "trackUnsubscribe") {
          setUnsubscribe(message.unsubscribed);
        }
      }
    } catch (error) {
      setNotification({
        type: "error",
        message: "Er ging iets mis met het ophalen van de data",
      });
    }
  });

  return (
    <>
      <AlertComponent notification={notification} />

      <div className="small-card p-2 bg-light d-grid gap-2 rounded shadow-lg">
        <p className="mb-0">Uitgeschreven : {unsubscribe}</p>
        <p className="mb-0">Online weergaven : {openedOnlineEmails}</p>
      </div>
    </>
  );
};

export default AnalyticsPanelCard;
