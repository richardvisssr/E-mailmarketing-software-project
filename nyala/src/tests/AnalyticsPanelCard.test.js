import { cleanup, render, screen } from "@testing-library/react";
import { createRoot } from "react-dom/client";
import { act } from "react-dom/test-utils";
import { useRouter } from "next/navigation";
import AnalyticsPanelCard from "../components/adminpanel/AnalyticsPanelCard";

jest.mock("next/navigation", () => ({
  useRouter: jest.fn(),
}));

global.fetch = jest.fn(() =>
  Promise.resolve({
    json: () => Promise.resolve({ data: "mockData" }),
  })
);

afterEach(() => {
  fetch.mockClear();
  cleanup();
});

it("should render the analytics panel card component", async () => {
  const div = document.createElement("div");
  const root = createRoot(div);

  await act(async () => {
    useRouter.mockImplementation(() => ({
      pathname: "",
      route: "",
      query: "",
      prefetch: jest.fn(),
    }));
    const mockAnalytics = { id: "mockId" };
    root.render(<AnalyticsPanelCard analytics={mockAnalytics} />);
  });

  expect(root.container).toMatchSnapshot();
});
