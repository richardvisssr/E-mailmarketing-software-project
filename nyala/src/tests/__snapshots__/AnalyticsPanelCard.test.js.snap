// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the analytics panel card component 1`] = `undefined`;
